File: pherix-ts/src/compensators/saas.ts
/**
 * SaaS-API compensators — the everyday agent tool calls.
 * Mirror of pherix/compensators/saas.py.
 *
 *   GitHub   createPr     → closePr          (open a PR → close it)
 *   GitHub   addLabel     → removeLabel       (label an issue → unlabel)
 *   Slack    postMessage  → deleteMessage     (post → delete)
 *   Stripe   createCustomer→ deleteCustomer    (create → delete)
 *   SendGrid addContact   → removeContact      (add to list → remove)
 *   Twilio   sendSms      → (GATE, no comp)    (cannot unsend an SMS)
 *   Jira     createIssue  → deleteIssue        (create → delete)
 *
 * Each clean pair reverses off a shared id in the args. `sendSms` has no honest
 * inverse (a delivered SMS is gone), so it gates exactly like `sendEmail`.
 */

import { tool, type ToolWrapper } from "../tools.js";

// --- GitHub ---------------------------------------------------------------

export interface GithubClient {
  createPr(repo: string, branch: string, title: string, body: string): unknown;
  closePr(repo: string, branch: string): unknown;
  addLabel(repo: string, issue: number, label: string): unknown;
  removeLabel(repo: string, issue: number, label: string): unknown;
}

export interface PrArgs extends Record<string, unknown> {
  repo: string;
  branch: string;
  title: string;
  body: string;
}

/** Register `createPr` and its left-inverse `closePr`. Closing (not merging) is
 *  the honest inverse of opening: an open PR has not yet changed the base
 *  branch, so closing it returns the repo to its pre-PR state. Reverses by
 *  `(repo, branch)`. */
export function registerGithubPr(
  client: GithubClient,
  resource = "github",
): { createPr: ToolWrapper<PrArgs, unknown>; closePr: ToolWrapper<PrArgs, unknown> } {
  const closePr = tool<PrArgs>(resource, (args: PrArgs) => client.closePr(args.repo, args.branch), {
    name: "closePr",
    reversible: false,
    injectsHandle: false,
  });

  const createPr = tool<PrArgs>(
    resource,
    (args: PrArgs) => client.createPr(args.repo, args.branch, args.title, args.body),
    { name: "createPr", reversible: false, injectsHandle: false, compensator: "closePr" },
  );

  return { createPr, closePr };
}

export interface LabelArgs extends Record<string, unknown> {
  repo: string;
  issue: number;
  label: string;
}

/** Register `addLabel` and its left-inverse `removeLabel`. Reverses by
 *  `(repo, issue, label)`. */
export function registerGithubLabel(
  client: GithubClient,
  resource = "github",
): { addLabel: ToolWrapper<LabelArgs, unknown>; removeLabel: ToolWrapper<LabelArgs, unknown> } {
  const removeLabel = tool<LabelArgs>(
    resource,
    (args: LabelArgs) => client.removeLabel(args.repo, args.issue, args.label),
    { name: "removeLabel", reversible: false, injectsHandle: false },
  );

  const addLabel = tool<LabelArgs>(
    resource,
    (args: LabelArgs) => client.addLabel(args.repo, args.issue, args.label),
    { name: "addLabel", reversible: false, injectsHandle: false, compensator: "removeLabel" },
  );

  return { addLabel, removeLabel };
}

// --- Slack ----------------------------------------------------------------

export interface SlackClient {
  postMessage(channel: string, ts: string, text: string): unknown;
  deleteMessage(channel: string, ts: string): unknown;
}

export interface MessageArgs extends Record<string, unknown> {
  channel: string;
  ts: string;
  text: string;
}

/** Register `postMessage` and its left-inverse `deleteMessage`. Reverses by
 *  `(channel, ts)` — the caller supplies the message timestamp as the
 *  idempotency key, matching Slack's own `chat.delete`. */
export function registerSlackMessage(
  client: SlackClient,
  resource = "slack",
): {
  postMessage: ToolWrapper<MessageArgs, unknown>;
  deleteMessage: ToolWrapper<MessageArgs, unknown>;
} {
  const deleteMessage = tool<MessageArgs>(
    resource,
    (args: MessageArgs) => client.deleteMessage(args.channel, args.ts),
    { name: "deleteMessage", reversible: false, injectsHandle: false },
  );

  const postMessage = tool<MessageArgs>(
    resource,
    (args: MessageArgs) => client.postMessage(args.channel, args.ts, args.text),
    { name: "postMessage", reversible: false, injectsHandle: false, compensator: "deleteMessage" },
  );

  return { postMessage, deleteMessage };
}

// --- Stripe ---------------------------------------------------------------

export interface StripeClient {
  createCustomer(customerId: string, email: string): unknown;
  deleteCustomer(customerId: string): unknown;
}

export interface CustomerArgs extends Record<string, unknown> {
  customerId: string;
  email: string;
}

/** Register `createCustomer` and its left-inverse `deleteCustomer`. Reverses by
 *  `customerId`. */
export function registerStripeCustomer(
  client: StripeClient,
  resource = "stripe",
): {
  createCustomer: ToolWrapper<CustomerArgs, unknown>;
  deleteCustomer: ToolWrapper<CustomerArgs, unknown>;
} {
  const deleteCustomer = tool<CustomerArgs>(
    resource,
    (args: CustomerArgs) => client.deleteCustomer(args.customerId),
    { name: "deleteCustomer", reversible: false, injectsHandle: false },
  );

  const createCustomer = tool<CustomerArgs>(
    resource,
    (args: CustomerArgs) => client.createCustomer(args.customerId, args.email),
    {
      name: "createCustomer",
      reversible: false,
      injectsHandle: false,
      compensator: "deleteCustomer",
    },
  );

  return { createCustomer, deleteCustomer };
}

// --- SendGrid -------------------------------------------------------------

export interface SendgridClient {
  addContact(listId: string, email: string): unknown;
  removeContact(listId: string, email: string): unknown;
}

export interface ContactArgs extends Record<string, unknown> {
  listId: string;
  email: string;
}

/** Register `addContact` and its left-inverse `removeContact`. Reverses by
 *  `(listId, email)`. Adding a contact to a list is cleanly reversible — note
 *  this is distinct from *sending* mail, which is not (see
 *  `registerSendEmailGate`). */
export function registerSendgridContact(
  client: SendgridClient,
  resource = "sendgrid",
): { addContact: ToolWrapper<ContactArgs, unknown>; removeContact: ToolWrapper<ContactArgs, unknown> } {
  const removeContact = tool<ContactArgs>(
    resource,
    (args: ContactArgs) => client.removeContact(args.listId, args.email),
    { name: "removeContact", reversible: false, injectsHandle: false },
  );

  const addContact = tool<ContactArgs>(
    resource,
    (args: ContactArgs) => client.addContact(args.listId, args.email),
    { name: "addContact", reversible: false, injectsHandle: false, compensator: "removeContact" },
  );

  return { addContact, removeContact };
}

// --- Twilio ---------------------------------------------------------------

export interface TwilioClient {
  sendSms(to: string, body: string): unknown;
}

export interface SmsArgs extends Record<string, unknown> {
  to: string;
  body: string;
}

/** Register `sendSms` with no compensator — it gates at commit. Like a
 *  delivered email, a delivered SMS has no honest inverse. Returns the action
 *  only. */
export function registerTwilioSmsGate(
  client: TwilioClient,
  resource = "twilio",
): ToolWrapper<SmsArgs, unknown> {
  return tool<SmsArgs>(resource, (args: SmsArgs) => client.sendSms(args.to, args.body), {
    name: "sendSms",
    reversible: false,
    injectsHandle: false,
  });
}

// --- Jira -----------------------------------------------------------------

export interface JiraClient {
  createIssue(issueKey: string, project: string, summary: string): unknown;
  deleteIssue(issueKey: string): unknown;
}

export interface IssueArgs extends Record<string, unknown> {
  issueKey: string;
  project: string;
  summary: string;
}

/** Register `createIssue` and its left-inverse `deleteIssue`. Reverses by
 *  `issueKey`. */
export function registerJiraCreateDeleteIssue(
  client: JiraClient,
  resource = "jira",
): { createIssue: ToolWrapper<IssueArgs, unknown>; deleteIssue: ToolWrapper<IssueArgs, unknown> } {
  const deleteIssue = tool<IssueArgs>(
    resource,
    (args: IssueArgs) => client.deleteIssue(args.issueKey),
    { name: "deleteIssue", reversible: false, injectsHandle: false },
  );

  const createIssue = tool<IssueArgs>(
    resource,
    (args: IssueArgs) => client.createIssue(args.issueKey, args.project, args.summary),
    { name: "createIssue", reversible: false, injectsHandle: false, compensator: "deleteIssue" },
  );

  return { createIssue, deleteIssue };
}

File: pherix/compensators/saas.py
"""SaaS-API compensators — the everyday agent tool calls.

These are the inverses an agent wiring up GitHub / Slack / Stripe /
SendGrid / Twilio / Jira reaches for first:

  GitHub   create_pr      → close_pr            (open a PR → close it)
  GitHub   add_label      → remove_label        (label an issue → unlabel)
  Slack    post_message   → delete_message      (post → delete)
  Stripe   create_customer→ delete_customer     (create → delete)
  SendGrid add_contact    → remove_contact      (add to list → remove)
  Twilio   send_sms       → (GATE, no comp)     (cannot unsend an SMS)
  Jira     create_issue   → delete_issue        (create → delete)

Each clean pair reverses off a shared id in the args. ``send_sms`` has no
honest inverse (a delivered SMS is gone), so it gates exactly like
``send_email`` — the catalog stays honest about what it cannot undo.
"""

from __future__ import annotations

from pherix.core.tools import tool


# --- GitHub ---------------------------------------------------------------


def register_github_pr(client, *, resource: str = "github"):
    """Register ``create_pr`` and its left-inverse ``close_pr``.

    ``client`` must expose::

        client.create_pr(repo, branch, title, body) -> object
        client.close_pr(repo, branch) -> object

    Reverses by ``(repo, branch)`` — closing the PR opened from that branch.
    Closing (not merging) is the honest inverse of opening: an open PR has
    not yet changed the base branch, so closing it returns the repo to its
    pre-PR state.
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def close_pr(repo, branch, title, body):
        return client.close_pr(repo, branch)

    @tool(
        resource=resource,
        reversible=False,
        injects_handle=False,
        compensator="close_pr",
    )
    def create_pr(repo, branch, title, body):
        return client.create_pr(repo, branch, title, body)

    return create_pr, close_pr


def register_github_label(client, *, resource: str = "github"):
    """Register ``add_label`` and its left-inverse ``remove_label``.

    ``client`` must expose::

        client.add_label(repo, issue, label) -> object
        client.remove_label(repo, issue, label) -> object

    Reverses by ``(repo, issue, label)``.
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def remove_label(repo, issue, label):
        return client.remove_label(repo, issue, label)

    @tool(
        resource=resource,
        reversible=False,
        injects_handle=False,
        compensator="remove_label",
    )
    def add_label(repo, issue, label):
        return client.add_label(repo, issue, label)

    return add_label, remove_label


# --- Slack ----------------------------------------------------------------


def register_slack_message(client, *, resource: str = "slack"):
    """Register ``post_message`` and its left-inverse ``delete_message``.

    ``client`` must expose::

        client.post_message(channel, ts, text) -> object
        client.delete_message(channel, ts) -> object

    Reverses by ``(channel, ts)``. The caller supplies the message timestamp
    ``ts`` as the idempotency key, so the action and its deletion share the
    handle the compensator receives — Slack's own ``chat.delete`` keys on
    exactly ``(channel, ts)``.
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def delete_message(channel, ts, text):
        return client.delete_message(channel, ts)

    @tool(
        resource=resource,
        reversible=False,
        injects_handle=False,
        compensator="delete_message",
    )
    def post_message(channel, ts, text):
        return client.post_message(channel, ts, text)

    return post_message, delete_message


# --- Stripe ---------------------------------------------------------------


def register_stripe_customer(client, *, resource: str = "stripe"):
    """Register ``create_customer`` and its left-inverse ``delete_customer``.

    ``client`` must expose::

        client.create_customer(customer_id, email) -> object
        client.delete_customer(customer_id) -> object

    Reverses by ``customer_id``.
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def delete_customer(customer_id, email):
        return client.delete_customer(customer_id)

    @tool(
        resource=resource,
        reversible=False,
        injects_handle=False,
        compensator="delete_customer",
    )
    def create_customer(customer_id, email):
        return client.create_customer(customer_id, email)

    return create_customer, delete_customer


# --- SendGrid -------------------------------------------------------------


def register_sendgrid_contact(client, *, resource: str = "sendgrid"):
    """Register ``add_contact`` and its left-inverse ``remove_contact``.

    ``client`` must expose::

        client.add_contact(list_id, email) -> object
        client.remove_contact(list_id, email) -> object

    Reverses by ``(list_id, email)``. Adding a contact to a marketing list
    is cleanly reversible — note this is distinct from *sending* mail to
    them, which is not (see ``identity.register_send_email_gate``).
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def remove_contact(list_id, email):
        return client.remove_contact(list_id, email)

    @tool(
        resource=resource,
        reversible=False,
        injects_handle=False,
        compensator="remove_contact",
    )
    def add_contact(list_id, email):
        return client.add_contact(list_id, email)

    return add_contact, remove_contact


# --- Twilio ---------------------------------------------------------------


def register_twilio_sms_gate(client, *, resource: str = "twilio"):
    """Register ``send_sms`` with **no compensator** — it gates at commit.

    ``client`` must expose ``client.send_sms(to, body) -> object``.

    Like a delivered email, a delivered SMS has no honest inverse, so the
    action stages and ``commit()`` blocks until ``approve_irreversible()``.
    Returns the action only.
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def send_sms(to, body):
        return client.send_sms(to, body)

    return send_sms


# --- Jira -----------------------------------------------------------------


def register_jira_create_delete_issue(client, *, resource: str = "jira"):
    """Register ``create_issue`` and its left-inverse ``delete_issue``.

    ``client`` must expose::

        client.create_issue(issue_key, project, summary) -> object
        client.delete_issue(issue_key) -> object

    Reverses by ``issue_key``.
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def delete_issue(issue_key, project, summary):
        return client.delete_issue(issue_key)

    @tool(
        resource=resource,
        reversible=False,
        injects_handle=False,
        compensator="delete_issue",
    )
    def create_issue(issue_key, project, summary):
        return client.create_issue(issue_key, project, summary)

    return create_issue, delete_issue
